### PR TITLE
first try for a color picker pop up

### PR DESCRIPTION
**Pull request by moinmoin-sh**
Here is a first try for a color picker popup. It utilises the color wheel as supported by the LVGL library. I increased the arc to make it easier touch which is not very responsive. The knob shows always the actual selected color.
It has 3 modes : Color, Saturation and Brightness.
You can cycle thru these modes with a long press on the center of the color wheel. The actual selected mode is displayed below the color wheel. Furthermore the resulting R, G ,B values are shown.
Double click on the center resets the selected mode.
No events are sent so far. This version is only meant as a showcase and ask how it feels before investing more effort.
colorpicker

### CODE REVIEW

**Comment by sumnerboy12** [comment]
Had a play - I like it - nice and easy to adjust. I am not sure we need the saturation control - although I am not an RGB expert.

I would be interested to see how a colour temp picker would work - I presume it would just be another ring. Need to think about how we configure it I guess. I.e. I have a load of CCT strips, so only need colour temp and brightness selectors. But other strips might be RGB or RGBCCT.

Seems like a very good start tho.

**Comment by austinscreations** [comment]
saturation no, not needed for our leds / flows

i'd say color wheel with brightness slider
then another 2 sliders. one for color temp, then one for brightness in relation to color temp. but changing any color temp or color temp brightness slider outputs current value, and then color wheel or rgb brightness outputs all variables needed

ideally rgb brightness color gradient matches current selected color, and ideally you'd have a way to push the newest color to teh display as well - should a different interface update the leds

**Comment by plains203** [comment]
I think Austin is right with his thoughts on that. The wheel for colour seems to work easily enough. The only issue I see with a slider is that it would also need to have an indication of what the brightness is doing and the colour temperature. It would be great if you could see the sliders at the same time as the wheel. Thoughts anyone?

**Comment by austinscreations** [comment]
just as potential reference i'll copy over this image from issue #2

178365563-9cf0a9ca-5ff6-4645-bcec-c9680eefd3f2

**Comment by moinmoin-sh** [comment]
What is the preferred units and range for the color temperature.
like 2000 ...  6000 K ?    or  `mireds `https://en.wikipedia.org/wiki/Mired 
I prefer something fixed. The backend then needs to transform it to the used range/units.

and brightness  ?   0 .. 100   or  0 .. 255

**Comment by sumnerboy12** [comment]
Brightness 1-100 i think, although could you re use the level start and stop config for consistency with the button level tiles?

Then it would be up to the user.

As for colour temp, I would vote for Kelvin but easy enough to convert in the back end.

**Comment by austinscreations** [comment]
i agree with ben

brightness 0-100

on mine i've been using 3000-6000 as my scale.

**Comment by moinmoin-sh** [comment]
This is a little further should include your comments above (almost), still no messaging.
The RGB values below the color wheel are the result from the color wheel x brightness (the right vertical bar)
Not sure if you really don't want the saturation ?  - that results in NO white (no shades of grey)
colorpicker2

**Comment by sumnerboy12** [comment]
How do you plan to support a CCT strip, i.e. just colour temp, no RGB. Just hide the colour wheel?

**Comment by austinscreations** [comment]
would essentially need 3 modes,
`RGB` + `RGBCCT` + `CCT`

**Comment by moinmoin-sh** [comment]
Yes, the screen could serve all 3 modes by hiding what is not needed. RGB and CCT are independent

**Comment by moinmoin-sh** [comment]
One more thing regarding how to use this.
As of now a short click on the tile opens the color picker, which means the short click is not available to control the connected light. Doesn't sound desirable to me.  Should the tile behavior changed in the way that the color picker is shown on long press so that the short press is free to switch the light?

**Comment by sumnerboy12** [comment]
Either that, or have an on/off button in the popup. I think I would
probably prefer your idea tho.

**Comment by austinscreations** [comment]
i liek idea of short press toggle, long opens menu

**Comment by moinmoin-sh** [comment]
That's all crap!  The latest sample I showed simply doesn't work. We can't ignore standards and have to support Color, Saturation, Brightness (HSV in terms of LVGL).

We could use the colorwheel as I showed in the first example, but the disadvantage is here that we have a limited resolution  RGB565 (16bit) because it's all bound to the resolution of our display.
As an alternative I could build a 3 slider goup with Color, Saturation and Brightness and make my own calculations based on RGB888 (24bit). That would be much more universal and would fit standards.
The Color bar/slider would look like this
grafik

**Comment by plains203** [comment]
Being as the LVGL library supports HSV rather than HSB like tasmota I think we have to go with something like you have mentioned above Moin. I also prefer the idea of sliders instead of a circle, I find it easier to accurately select the colour you are after.

**Comment by austinscreations** [comment]
to add frustration to the mix i think neopixels can use HSV and RGB

**Comment by plains203** [comment]
> to add frustration to the mix i think neopixels can use HSV and RGB

Indeed but we need to work with what the screen will do and translate it to the lights. 
Have a quick read of https://docs.lvgl.io/latest/en/html/widgets/cpicker.html

**Comment by austinscreations** [comment]
thats the picker i think openHASP was using seemed to work well

**Comment by moinmoin-sh** [comment]
I hear you saying "let's do it the openHASP way" ?

**Comment by moinmoin-sh** [comment]
Sorry Austin,I wasn't clear in mylast post. I thought this PR/discussion is all in context of a color picker only, not general.
Anyway, I decided to use the LVGL colorpicker as is. The color related mqtt interface (stat/ cmnd/) wil be based on RGB which is well known everywhere.

**Comment by moinmoin-sh** [comment]
The latest commit has a fully working color picker
- the colorPicker popup is opened with a long press on the tile
- only the RGBCCT function is available
- the color wheel works as described in the first post
- all values of the colorPicker popup are sent on any change
- all values can be preset. Can be done at any time before the popup is opened, should be avoided when the popup is active.
- ColorPickerEvent :
```
{"screen":<number>, "tile":<number>, "style":"<style>", "type":"colorPicker","event":"change",
"State":{"colorRGB":{"red":<number>,"green":<number>,"blue":<number>},"colorTemperature":<value>,"brightness":<value>}}
```
-Set values for colorPicker 
```
{"tiles": [{"screen":<number>, "tile":<number>, "colorPicker":{"colorRGB":{"red":<number>,"green":<number>,"blue":<number>},
"colorTemperature":<number>,"brightness":<number>}}]} 
```

**Comment by moinmoin-sh** [comment]
How would you like to enable the RGB-only or CCT-only version of the colorPicker?

a) as seperate styles like `colorPickerAll`, `colorPickerRGB`, `colorPickerCCT` 
b) or set later with something like `{"tiles": [{"screen":<number>, "tile":<number>, “colorPickerHide”: "rgb"|"cct"}]}`

**Comment by sumnerboy12** [comment]
Separate styles is probably more consistent with the other styles, like button levels etc.

**Comment by moinmoin-sh** [comment]
supports now styles `colorPickerRgb`, `colorPickerCct`, `colorPickerRgbCct`
rgb
cct
rgb_cct

File: src/main.cpp [hunk 1]
```diff
@@ -224,7 +228,9 @@ void initStyleLut(void)
   styleLut[TS_BUTTON_LEFT_RIGHT] = {TS_BUTTON_LEFT_RIGHT, "buttonLeftRight"};
   styleLut[TS_BUTTON_PREV_NEXT] = {TS_BUTTON_PREV_NEXT, "buttonPrevNext"};
   styleLut[TS_INDICATOR] = {TS_INDICATOR, "indicator"};
-  styleLut[TS_COLOR_PICKER] = {TS_COLOR_PICKER, "colorPicker"};
+  styleLut[TS_COLOR_PICKER_RGB_CCT] = {TS_COLOR_PICKER_RGB_CCT, "colorPickerRgbCct"};
```
Comment (sumnerboy12) [inline src/main.cpp:231]: Sorry to be pedantic, but I think the `RGB` and `CCT` abbreviations should be consistent - i.e. here and in the status payload are currently different. Don't mind which way you go, just ensure they are the same throughout.

Probably easier to change the status payload key to `colorRgb`?

Comment (sumnerboy12) [reply src/main.cpp:231]: Fix proposed [here](https://github.com/austinscreations/OXRS-AC-TP32-ESP-FW/pull/26)

**Pull request merged**